src/coreclr/src/vm/codeversion.cpp

```diff
@@ -1670,6 +1671,14 @@ PCODE CodeVersionManager::PublishVersionableCodeIfNecessary(
         {
             PrepareCodeConfigBuffer configBuffer(activeVersion);
             PrepareCodeConfig *config = configBuffer.GetConfig();
+
+#ifdef FEATURE_TIERED_COMPILATION
+            // If the caller is in preemptive mode, then
+            // Tiered Compilation isn't supported. (e.g. NativeCallable)
+            if (callerGCMode == MethodDesc::CallerGCMode_Preempt)
+                config->SetJitSwitchedToOptimized();
```

Comment:
There are other places in the system that call PrepareCode. Does this need to be pushed into `PrepareCode`?

Reply:
As in the propagation of the caller's GC mode?

Reply:
There appears to be at least one relevant call here (error path), but there may be others in the future:
https://github.com/dotnet/runtime/commit/c78f4a4288aeee32938e0f96535e52d0978bfb69#diff-b06261f7127af3268df1f78173d20b34R1659

Wonder if it would be safer to put the enum value into `PrepareCodeConfig` and check for that after looking up R2R'ed code instead of using `JitSwitchedToOptimized`. If that value is Unknown (default value when not explicitly set, this would hopefully cover all other paths) then it can do the expensive lookup.

Reply:
And that value could then be reused on the JIT path as well to avoid the expensive lookup there

Reply:
So enrich `PrepareCodeConfig`? That might make the enum build issue a bit easier to address. Let me try that.

Reply:
Well that isn't going to address the build issues. This is an absolute mess.

Reply:
> that value could then be reused on the JIT path

I think we should always do the attribute lookup on the JIT path to get the right code generated. If we do not do that, I think the detection of incorrect use is not going to work well.